### PR TITLE
Fix recurring payments.

### PR DESCRIPTION
**Pull request by mworrell**
Fixes handling pushes of new recurring payments.
Removes storage of subscription and customer id in person record.

Mantis issue 0030141

See also https://github.com/driebit/mod_payment/pull/2

### CODE REVIEW

File: models/m_payment_mollie_api.erl [hunk 1]
```diff
@@ -137,65 +147,240 @@ payment_url(MollieId) ->
...
+%% @doc Pull the payment status from Mollie.
+-spec payment_sync(binary() | integer(), binary(), z:context()) -> ok | {error, notfound|term()}.
+payment_sync(PaymentNrOrId, ExtId, Context) ->
+    case m_payment:get(PaymentNrOrId, Context) of
+        {ok, Payment} ->
+            case proplists:get_value(psp_module, Payment) of
+                mod_payment_mollie ->
+                    % Fetch the status from Mollie
+                    PaymentId = proplists:get_value(id, Payment),
+                    lager:info("Payment PSP Mollie webhook call for payment #~p", [PaymentId]),
```
Comment (robvandenbogaard) [inline models/m_payment_mollie_api.erl:247]: Is this a webhook (callback) call or just a request to the Mollie payment api?

Comment (mworrell) [reply models/m_payment_mollie_api.erl:247]: This routine is (at the moment) only called in response to a web hook call by Mollie.

File: models/m_payment_mollie_api.erl [hunk 2]
```diff
@@ -137,65 +147,240 @@ payment_url(MollieId) ->
...
+        <<"description">> := Description
+    } = JSON,
+    % Create a new payment, referring to the PaymentId
+    Currency = proplists:get_value(currency, Payment),
+    AmountNr = z_convert:to_float(Amount),
+    DateTime = z_convert:to_datetime( status_date(JSON) ),
+    case m_payment:get_by_psp(mod_payment_mollie, ExtId, Context) of
+        {ok, Payment} ->
+            PaymentId = proplists:get_value(id, Payment),
+            update_payment_status(PaymentId, Status, DateTime, Context);
```
Comment (robvandenbogaard) [inline models/m_payment_mollie_api.erl:303]: Should `m_payment:update_pdp_handler` also be called here, as on line 313?

File: models/m_payment_mollie_api.erl [hunk 3]
```diff
@@ -260,129 +443,299 @@ api_key(Context) ->
...
-        end
+    case list_subscriptions(UserId, Context) of
+        {ok, Subs} ->
+            lists:any(fun is_subscription_active/1, Subs);
+        {error, _} ->
+            false
     end.
 
+is_subscription_active(#{ end_date := undefined }) -> false;
+is_subscription_active(#{ end_date := _}) -> true.
```
Comment (robvandenbogaard) [inline models/m_payment_mollie_api.erl]: Shouldn't this be the other way round? If there is no end-date at which the sub was terminated, it should be considered active - if I'm correct I've seen that interpretation somewhere else in the code above.

File: models/m_payment_mollie_api.erl (hunk 2 above)
Comment (mworrell) [reply models/m_payment_mollie_api.erl:303]: Here the PSP is already correctly registered, so we don't need to update those fields.
On line 313 we add a new payment, and need to update that with the correct PSP.

File: models/m_payment_mollie_api.erl [hunk 4]
```diff
@@ -260,129 +443,299 @@ api_key(Context) ->
...
+first_valid_custid([ CustomerId | CustIds ], Context) ->
+    case api_call(get, "customers/" ++ z_convert:to_list(CustomerId), [], Context) of
+        {ok, _Cust} ->
+            {ok, CustomerId};
+        {error, 404} ->
+            first_valid_custid(CustIds, Context);
+        {error, 410} ->
+            first_valid_custid(CustIds, Context);
+        {error, _} = Error ->
             Error
```
Comment (robvandenbogaard) [inline models/m_payment_mollie_api.erl:707]: Nicely robust, to be able to fall back to older customer ids.

File: models/m_payment_mollie_api.erl (hunk 3 above)
Comment (mworrell) [reply models/m_payment_mollie_api.erl]: Good one!

Let me check this and the code calling this.

Comment (mworrell) [reply models/m_payment_mollie_api.erl]: You are right - I pushed a fix.

**Pull request merged**